File: NextApp/Android/core/model/src/main/kotlin/jp/komehyappyo/member/next/core/model/Diary.kt
package jp.komehyappyo.member.next.core.model

import java.time.Instant
import java.util.UUID

enum class DiaryMood {
    VeryGood,
    Good,
    Neutral,
    SlightlyBad,
    Bad,
}

data class Diary(
    val id: UUID = UUID.randomUUID(),
    val userId: String,
    val title: String,
    val body: String = "",
    val mood: DiaryMood = DiaryMood.Neutral,
    val photoUrls: List<String> = emptyList(),
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now(),
) {
    fun validated(now: Instant = Instant.now()): Diary {
        require(title.trim().isNotEmpty()) { "タイトルを入力してください。" }
        require(photoUrls.size <= MAXIMUM_PHOTO_COUNT) {
            "写真は${MAXIMUM_PHOTO_COUNT}枚まで登録できます。"
        }
        return copy(
            title = title.trim(),
            body = body.trim(),
            updatedAt = now,
        )
    }

    companion object {
        const val MAXIMUM_PHOTO_COUNT = 5
    }
}

File: NextApp/iOS/OrgPortalNext/Core/Model/Diary.swift
import Foundation

public enum DiaryMood: String, CaseIterable, Codable, Sendable {
    case veryGood
    case good
    case neutral
    case slightlyBad
    case bad

    public var localizationKey: String {
        switch self {
        case .veryGood: "diary.mood.very_good"
        case .good: "diary.mood.good"
        case .neutral: "diary.mood.neutral"
        case .slightlyBad: "diary.mood.slightly_bad"
        case .bad: "diary.mood.bad"
        }
    }
}

public struct Diary: Identifiable, Equatable, Codable, Sendable {
    public static let maximumPhotoCount = 5

    public var id: UUID
    public var userId: String
    public var title: String
    public var body: String
    public var mood: DiaryMood
    public var photoUrls: [String]
    public var createdAt: Date
    public var updatedAt: Date

    public init(
        id: UUID = UUID(),
        userId: String,
        title: String,
        body: String = "",
        mood: DiaryMood = .neutral,
        photoUrls: [String] = [],
        createdAt: Date = .now,
        updatedAt: Date = .now
    ) {
        self.id = id
        self.userId = userId
        self.title = title
        self.body = body
        self.mood = mood
        self.photoUrls = photoUrls
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    public func validated(now: Date = .now) throws -> Diary {
        var result = self
        result.title = title.trimmingCharacters(in: .whitespacesAndNewlines)
        result.body = body.trimmingCharacters(in: .whitespacesAndNewlines)
        result.updatedAt = now

        guard !result.title.isEmpty else {
            throw DiaryValidationError.titleRequired
        }
        guard result.photoUrls.count <= Self.maximumPhotoCount else {
            throw DiaryValidationError.tooManyPhotos
        }
        return result
    }
}

public enum DiaryValidationError: Error, Equatable, Sendable {
    case titleRequired
    case tooManyPhotos

    public var localizationKey: String {
        switch self {
        case .titleRequired: "error.diary.title_required"
        case .tooManyPhotos: "error.diary.too_many_photos"
        }
    }
}
